### PR TITLE
[SPARK-50130][SQL][PYTHON] Add DataFrame APIs for scalar and exists subqueries

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Adds the following DataFrame APIs for subqueries to Spark Classic.

- `scalar()`
- `exists()`

Also, add `outer()` to `Column` to specify outer references.

#### Examples:

For the following tables `l` and `r`:

```py
>>> spark.table("l").printSchema()
root
 |-- a: long (nullable = true)
 |-- b: double (nullable = true)

>>> spark.table("r").printSchema()
root
 |-- c: long (nullable = true)
 |-- d: double (nullable = true)
```

```py
from pyspark.sql import functions as sf

# select * from l where b < (select max(d) from r where a = c)
spark.table("l").where(
    sf.col("b")
    < (
        spark.table("r")
        .where(sf.col("a").outer() == sf.col("c"))
        .select(sf.max("d"))
        .scalar()
    )
)

# select a, (select sum(b) from l l2 where l2.a = l1.a) sum_b from l l1
spark.table("l").select(
    "a",
    (
        spark.table("l")
        .where(sf.col("a") == sf.col("a").outer())
        .select(sf.sum("b"))
        .scalar()
        .alias("sum_b")
    ),
)

# select * from l where exists (select * from r where l.a = r.c)
spark.table("l").where(
    spark.table("r").where(sf.col("a").outer() == sf.col("c")).exists()
)
```

### Why are the changes needed?

Subquery APIs are missing in DataFrame API.

### Does this PR introduce _any_ user-facing change?

Yes, new DataFrame APIs for subqueries will be available.

### How was this patch tested?

Added the related tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @allisonwang-db @cloud-fan @dtenedor

**Comment by ueshin** [comment]
I'll update documents here, and work on Spark Connect in a separate PR.

File: sql/api/src/main/scala/org/apache/spark/sql/Column.scala [hunk 1]
```diff
@@ -1382,6 +1382,17 @@ class Column(val node: ColumnNode) extends Logging {
...
 
+  /**
+   * TODO: Add documentation
+   * @return
+   */
+  def outer(): Column = withOrigin {
+    node match {
+      case attr: UnresolvedAttribute if !attr.isMetadataColumn =>
+        Column(LazyOuterReference(attr.nameParts, attr.planId))
+      // TODO: other cases
```
Comment (ueshin) [inline sql/api/src/main/scala/org/apache/spark/sql/Column.scala:1393]: Is it ok to use metadata columns as outer?
Do we have other cases we can support `outer()`?

**Review by HyukjinKwon — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -6476,6 +6476,153 @@ def transpose(self, indexColumn: Optional["ColumnOrName"] = None) -> "DataFrame"
...
+        --------
+        Setup a sample DataFrame.
+
+        >>> data = [
+        ...     (1, "Alice", 45000, 101), (2, "Bob", 54000, 101), (3, "Charlie", 29000, 102),
+        ...     (4, "David", 61000, 102), (5, "Eve", 48000, 101),
+        ... ]
+        >>> employees = spark.createDataFrame(data, ["id", "name", "salary", "department_id"])
+
+        Example 1: Filter for employees with salary greater than the average salary.
```
Comment (allisonwang-db) [inline python/pyspark/sql/dataframe.py:6506]: Can we also mention here this is a non-correlated example?

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -6476,6 +6476,153 @@ def transpose(self, indexColumn: Optional["ColumnOrName"] = None) -> "DataFrame"
...
+        ... ).select("name", "salary", "department_id").show()
+        +-----+------+-------------+
+        | name|salary|department_id|
+        +-----+------+-------------+
+        |  Bob| 54000|          101|
+        |David| 61000|          102|
+        |  Eve| 48000|          101|
+        +-----+------+-------------+
+
+        Example 2: Filter for employees with salary greater than the average salary in their
```
Comment (allisonwang-db) [inline python/pyspark/sql/dataframe.py:6520]: ditto, let's mention that this is a correlated subquery example

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -6476,6 +6476,153 @@ def transpose(self, indexColumn: Optional["ColumnOrName"] = None) -> "DataFrame"
...
+        >>> customers = spark.createDataFrame(
+        ...     data_customers, ["customer_id", "customer_name", "country"])
+        >>> orders = spark.createDataFrame(
+        ...     data_orders, ["order_id", "customer_id", "order_date", "total_amount"])
+
+        Example 1: Filter for customers who have placed at least one order.
+
+        >>> from pyspark.sql import functions as sf
+        >>> customers.where(
+        ...     orders.where(sf.col("customer_id") == sf.col("customer_id").outer()).exists()
```
Comment (allisonwang-db) [inline python/pyspark/sql/dataframe.py:6597]: Does it work with NOT EXISTS?

File: sql/api/src/main/scala/org/apache/spark/sql/internal/columnNodes.scala [hunk 5]
```diff
@@ -167,6 +167,16 @@ private[sql] object UnresolvedAttribute {
     apply(unparsedIdentifier, None, false, CurrentOrigin.get)
 }
 
+private[sql] case class LazyOuterReference(
```
Comment (allisonwang-db) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/columnNodes.scala:170]: Let's add a javadoc here explaining this node?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -2482,9 +2482,16 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
      */
     private def resolveSubQuery(
         e: SubqueryExpression,
-        outer: LogicalPlan)(
+        outer: LogicalPlan,
+        hasExplicitOuterRefs: Boolean = false)(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2486]: nit: let's add this new parameter to the doc above?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -245,13 +262,17 @@ class Dataset[T] private[sql](
   }
 
   @transient private[sql] val logicalPlan: LogicalPlan = {
-    val plan = queryExecution.commandExecuted
-    if (sparkSession.sessionState.conf.getConf(SQLConf.FAIL_AMBIGUOUS_SELF_JOIN_ENABLED)) {
-      val dsIds = plan.getTagValue(Dataset.DATASET_ID_TAG).getOrElse(new HashSet[Long])
-      dsIds.add(id)
-      plan.setTagValue(Dataset.DATASET_ID_TAG, dsIds)
+    if (queryExecution.isLazyAnalysis) {
+      queryExecution.logical
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:266]: cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 8]
```diff
@@ -106,8 +111,13 @@ private[sql] object Dataset {
     sparkSession.withActive {
       val qe = new QueryExecution(
         sparkSession, logicalPlan, shuffleCleanupMode = shuffleCleanupMode)
-      qe.assertAnalyzed()
-      new Dataset[Row](qe, RowEncoder.encoderFor(qe.analyzed.schema))
+      val encoder = if (qe.isLazyAnalysis) {
+        RowEncoder.encoderFor(new StructType())
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:115]: I wonder if we can add a test to trigger this case

**Review by allisonwang-db — COMMENTED** [review]
Awesome feature! Looks good overall!

File: python/pyspark/sql/dataframe.py (hunk 4 above)
Comment (ueshin) [reply python/pyspark/sql/dataframe.py:6597]: Yes, let me add the example.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 8 above)
Comment (ueshin) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:115]: Every test using outer references will pass this case?

**Review by xinrong-meng — APPROVED** [review]

**Comment by xinrong-meng** [comment]
[Run / Build modules: pyspark-mllib, pyspark-ml, pyspark-ml-connect](https://github.com/ueshin/apache-spark/actions/runs/11714014185/job/32627949792#logs) seems to have been canceled for some reason. Rerun the jobs.

**Comment by ueshin** [comment]
The failing tests are not related to this PR.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/api/src/main/scala/org/apache/spark/sql/internal/columnNodes.scala [hunk 9]
```diff
@@ -167,6 +167,24 @@ private[sql] object UnresolvedAttribute {
...
+ *   id of the plan (Dataframe) that produces the attribute.
+ */
+private[sql] case class LazyOuterReference(
+    nameParts: Seq[String],
+    planId: Option[Long] = None,
+    override val origin: Origin = CurrentOrigin.get)
+    extends ColumnNode {
+  override private[internal] def normalize(): LazyOuterReference =
+    copy(planId = None, origin = NO_ORIGIN)
+  override def sql: String = nameParts.map(n => if (n.contains(".")) s"`$n`" else n).mkString(".")
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/internal/columnNodes.scala:185]: let's be consistent with other parts in spark SQL and use `QuotingUtils.quoteIfNeeded`

File: sql/api/src/main/scala/org/apache/spark/sql/Column.scala [hunk 10]
```diff
@@ -1382,6 +1382,21 @@ class Column(val node: ColumnNode) extends Logging {
...
+   *
+   * @group subquery
+   * @since 4.0.0
+   */
+  def outer(): Column = withOrigin {
+    node match {
+      case attr: UnresolvedAttribute if !attr.isMetadataColumn =>
+        Column(LazyOuterReference(attr.nameParts, attr.planId))
+      case _ =>
+        throw new IllegalArgumentException(
```
Comment (cloud-fan) [inline sql/api/src/main/scala/org/apache/spark/sql/Column.scala:1396]: we should throw an internal error.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 11]
```diff
@@ -95,9 +95,14 @@ private[sql] object Dataset {
   def ofRows(sparkSession: SparkSession, logicalPlan: LogicalPlan): DataFrame =
     sparkSession.withActive {
       val qe = sparkSession.sessionState.executePlan(logicalPlan)
-      qe.assertAnalyzed()
-      new Dataset[Row](qe, RowEncoder.encoderFor(qe.analyzed.schema))
-  }
+      val encoder = if (qe.isLazyAnalysis) {
+        RowEncoder.encoderFor(new StructType())
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:99]: @ueshin this breaks collect (and other operations) on these dataframes. The alternative would be to defer the construction of the encoder until it is needed.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:265]: What happens if the Dataset is a command?

File: sql/api/src/main/scala/org/apache/spark/sql/Column.scala (hunk 10 above)
Comment (ueshin) [reply sql/api/src/main/scala/org/apache/spark/sql/Column.scala:1396]: This is a user-facing API, so I feel throwing an internal error is weird. I'm ok to change the exception type to other user-facing exception, though.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 11 above)
Comment (ueshin) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:99]: @hvanhovell sure, submitted #48829.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:265]: you can't put a command inside a subquery expression (neither in SQL).

**Pull request closed without merging**